Drop CString::dataAsUInt8Ptr() in favor of span()

cdumez opened the pull request:
#### 27853e7d160e2e47c96e2e7da0f0f5a152c0a418
<pre>
Drop CString::dataAsUInt8Ptr() in favor of span()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273804">https://bugs.webkit.org/show_bug.cgi?id=273804</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorConnectionClient.cpp:
(Inspector::RemoteInspectorConnectionClient::send):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorConnectionClient.h:
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
(Inspector::MessageParser::createMessage):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.h:
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocketEndpoint.cpp:
(Inspector::RemoteInspectorSocketEndpoint::send):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocketEndpoint.h:
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL):
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder&lt;CString&gt;::encodeForPersistence):
* Source/WTF/wtf/text/CString.cpp:
(WTF::operator==):
* Source/WTF/wtf/text/CString.h:
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::showText):
* Source/WebCore/platform/graphics/avfoundation/cf/CDMSessionAVFoundationCF.cpp:
(WebCore::CDMSessionAVFoundationCF::generateKeyRequest):
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.h:
(WebCore::httpHeaderValueUsingSuitableEncoding):
* Source/WebDriver/socket/SessionHostSocket.cpp:
(WebDriver::SessionHost::sendWebInspectorEvent):
* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCopyCFURL):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorClient.cpp:
(WebKit::RemoteInspectorClient::sendWebInspectorEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityTreeData):

Canonical link: <a href="https://commits.webkit.org/278470@main">https://commits.webkit.org/278470@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/015908b936d194e0ec8fd624737a1237d10e04b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53912 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1344 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36210 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/994 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52752 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22406 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/885 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9094 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/43989 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46971 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/947 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55502 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50156 "Found 1 new JSC binary failure: testapi (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25755 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/27012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43772 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27880 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/57631 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7335 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26744 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11855 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
```diff
@@ -46,16 +46,16 @@ MessageParser::MessageParser(Function<void(Vector<uint8_t>&&)>&& listener)
+Vector<uint8_t> MessageParser::createMessage(std::span<const uint8_t> data)
 {
-    if (!data || !size || size > UINT_MAX)
+    if (!data.data() || data.empty() || data.size() > UINT_MAX)
```
Should not be `UINT_MAX` here because `htonl` takes `uint32_t` and it's unnecessary to code in the assumption that `unsigned` is `uint32_t`. So we should use `std::numeric_limits<uint32_t>::max()`.

Null check of data pointer seems unneeded.

darinadler commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
```diff
@@ -46,16 +46,16 @@ MessageParser::MessageParser(Function<void(Vector<uint8_t>&&)>&& listener)
-    auto messageBuffer = Vector<uint8_t>(size + sizeof(uint32_t));
-    uint32_t uintSize = static_cast<uint32_t>(size);
+    auto messageBuffer = Vector<uint8_t>(data.size() + sizeof(uint32_t));
+    uint32_t uintSize = static_cast<uint32_t>(data.size());
```
The `static_cast` here is unneeded. And this local variable is unneeded.

darinadler commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
```diff
@@ -46,16 +46,16 @@ MessageParser::MessageParser(Function<void(Vector<uint8_t>&&)>&& listener)
     uint32_t nboSize = htonl(uintSize);
     memcpy(&messageBuffer[0], &nboSize, sizeof(uint32_t));
-    memcpy(&messageBuffer[sizeof(uint32_t)], data, uintSize);
+    memcpy(&messageBuffer[sizeof(uint32_t)], data.data(), uintSize);
```
This can just be `data.size()`.

darinadler commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
```diff
@@ -46,16 +46,16 @@ MessageParser::MessageParser(Function<void(Vector<uint8_t>&&)>&& listener)
-    uint32_t uintSize = static_cast<uint32_t>(size);
+    auto messageBuffer = Vector<uint8_t>(data.size() + sizeof(uint32_t));
+    uint32_t uintSize = static_cast<uint32_t>(data.size());
     uint32_t nboSize = htonl(uintSize);
```
This can just be `data.size()`.

darinadler commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
```diff
@@ -81,7 +81,7 @@ void RemoteInspector::sendWebInspectorEvent(const String& event)
         return;
 
     const CString message = event.utf8();
```
Should not use this local variable any more.

darinadler commented on Source/WebCore/platform/graphics/avfoundation/cf/CDMSessionAVFoundationCF.cpp:
```diff
@@ -67,8 +67,9 @@ RefPtr<Uint8Array> CDMSessionAVFoundationCF::generateKeyRequest(const String&, U
-    CFDataAppendBytes(assetID.get(), assetStr.dataAsUInt8Ptr(), assetStr.length());
+    auto assetStrSpan = assetStr.span();
+    auto assetID = adoptCF(CFDataCreateMutable(kCFAllocatorDefault, assetStrSpan.size()));
+    CFDataAppendBytes(assetID.get(), assetStrSpan.data(), assetStrSpan.size());
```
Seems like this should use a function that creates a `CFData` from a `span`, one that we already have. Seems highly unlikely this needs to be mutable. Probably same for the certificate above.

darinadler commented on Source/WebCore/platform/network/cf/ResourceRequestCFNet.h:
```diff
@@ -120,8 +120,9 @@ inline RetainPtr<CFStringRef> httpHeaderValueUsingSuitableEncoding(HTTPHeaderMap
     if (header.keyAsHTTPHeaderName && *header.keyAsHTTPHeaderName == HTTPHeaderName::LastEventID && !header.value.containsOnlyASCII()) {
         auto utf8Value = header.value.utf8();
+        auto utf8ValueSpan = utf8Value.span();
         // Constructing a string with the UTF-8 bytes but claiming that it’s Latin-1 is the way to get CFNetwork to put those UTF-8 bytes on the wire.
```
This seems like a super strange workaround for an issue in CFNetwork. I wonder if this can ever be rectified.

darinadler commented on Source/WebDriver/socket/HTTPServerSocket.cpp:
```diff
@@ -108,7 +108,7 @@ void HTTPRequestHandler::sendResponse(HTTPRequestHandler::Response&& response)
 {
     auto& endpoint = RemoteInspectorSocketEndpoint::singleton();
     auto packet = packHTTPMessage(WTFMove(response));
```
We don't need this local variable any more but we can keep it if we think it adds clarity.

darinadler commented on Source/WebDriver/socket/SessionHostSocket.cpp:
```diff
@@ -51,8 +51,8 @@ void SessionHost::sendWebInspectorEvent(const String& event)
 
-    const CString message = event.utf8();
-    send(m_clientID.value(), message.dataAsUInt8Ptr(), message.length());
+    auto message = event.utf8();
```
Would be better without this local variable.

darinadler commented on Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorClient.cpp:
```diff
@@ -126,7 +126,7 @@ void RemoteInspectorClient::sendWebInspectorEvent(const String& event)
     ASSERT(isMainRunLoop());
     ASSERT(m_connectionID);
     auto message = event.utf8();
```
Would be better without this local variable.

darinadler commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4552,8 +4552,12 @@ void WebPage::getAccessibilityTreeData(CompletionHandler<void(const std::optiona
-        CFWriteStreamWrite(stream.get(), treeData->isolatedTree.utf8().dataAsUInt8Ptr(), treeData->isolatedTree.utf8().length());
+        auto liveTreeUTF8 = treeData->liveTree.utf8();
+        auto liveTreeUTF8Span = liveTreeUTF8.span();
+        CFWriteStreamWrite(stream.get(), liveTreeUTF8Span.data(), liveTreeUTF8Span.size());
```
I would probably make a helper function or lambda to avoid wiring this three line thing out twice.

cdumez committed:
Drop CString::dataAsUInt8Ptr() in favor of span()
https://bugs.webkit.org/show_bug.cgi?id=273804

Reviewed by Darin Adler.

* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorConnectionClient.cpp:
(Inspector::RemoteInspectorConnectionClient::send):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorConnectionClient.h:
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.cpp:
(Inspector::MessageParser::createMessage):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorMessageParser.h:
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocketEndpoint.cpp:
(Inspector::RemoteInspectorSocketEndpoint::send):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocketEndpoint.h:
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL):
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder<CString>::encodeForPersistence):
* Source/WTF/wtf/text/CString.cpp:
(WTF::operator==):
* Source/WTF/wtf/text/CString.h:
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::showText):
* Source/WebCore/platform/graphics/avfoundation/cf/CDMSessionAVFoundationCF.cpp:
(WebCore::CDMSessionAVFoundationCF::generateKeyRequest):
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.h:
(WebCore::httpHeaderValueUsingSuitableEncoding):
* Source/WebDriver/socket/SessionHostSocket.cpp:
(WebDriver::SessionHost::sendWebInspectorEvent):
* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCopyCFURL):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorClient.cpp:
(WebKit::RemoteInspectorClient::sendWebInspectorEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityTreeData):

Canonical link: https://commits.webkit.org/278470@main
